src/type.c

```diff
@@ -72,6 +48,31 @@ SEXP vec_type(SEXP x) {
   return vec_type_rec(x, true);
 }
 
+
+SEXP vec_type_finalise_rec(SEXP x, bool dispatch) {
+  if (OBJECT(x)) {
```

Comment:
I don't quite understand the logic between how you've broken up the cases here. I would expect to see `unspecified`, `data.frame`, and `partial` classes more prominently

Reply:
`unspecified` and `partial` are not part of the vctrs enum so are/will be handled in the `if (OBJECT(x))` branch. Partial types are kept in R for now but will be included here in the future.